### ISSUE TITLE
assertion failed when testing examples/server.rs

### ISSUE THREAD

**Issue by bbigras** [snapview/tokio-tungstenite#7]
To reproduce, use the [Simple WebSocket Client](https://chrome.google.com/webstore/detail/simple-websocket-client/pfdhoblngboilpfeibdedpjgfnlcodoo) chrome extension and connect then disconnect.

```
> RUST_BACKTRACE=full cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/test-tung-server`
Listening on: 127.0.0.1:8080
New WebSocket connection: 127.0.0.1:50700
thread 'main' panicked at 'assertion failed: `(left == right)` (left: `1`, right: `0`)', /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/frame/frame.rs:26
stack backtrace:
   0:     0x55667b188153 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::h0c49f46a3545f908
                               at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x55667b1843b4 - std::sys_common::backtrace::_print::hcef39a9816714c4c
                               at /checkout/src/libstd/sys_common/backtrace.rs:71
   2:     0x55667b18a287 - std::panicking::default_hook::{{closure}}::h7c3c94835e02f846
                               at /checkout/src/libstd/sys_common/backtrace.rs:60
                               at /checkout/src/libstd/panicking.rs:355
   3:     0x55667b189e0b - std::panicking::default_hook::h0bf7bc3112fb107d
                               at /checkout/src/libstd/panicking.rs:371
   4:     0x55667b18a6fb - std::panicking::rust_panic_with_hook::ha27630c950090fec
                               at /checkout/src/libstd/panicking.rs:549
   5:     0x55667b18a5d4 - std::panicking::begin_panic::heb97fa3158b71158
                               at /checkout/src/libstd/panicking.rs:511
   6:     0x55667b18a509 - std::panicking::begin_panic_fmt::h8144403278d84748
                               at /checkout/src/libstd/panicking.rs:495
   7:     0x55667b0f723e - tungstenite::protocol::frame::frame::apply_mask_aligned32::h2e9587b64b443a55
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/frame/frame.rs:26
   8:     0x55667b09bfa6 - tungstenite::protocol::frame::frame::Frame::remove_mask::{{closure}}::h464b933611967a31
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/frame/frame.rs:224
   9:     0x55667b0a3fc5 - <core::option::Option<T>>::and_then::hc6cbc7d04a997358
                               at /checkout/src/libcore/option.rs:587
  10:     0x55667b09bf35 - tungstenite::protocol::frame::frame::Frame::remove_mask::h7d7a272641f1a4d0
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/frame/frame.rs:221
  11:     0x55667b0c71e6 - <tungstenite::protocol::WebSocket<Stream>>::read_message_frame::hdce73b2af0b2291f
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/mod.rs:197
  12:     0x55667b0c4939 - <tungstenite::protocol::WebSocket<Stream>>::read_message::he6e38db2dd59fe22
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tungstenite-0.2.0/src/protocol/mod.rs:93
  13:     0x55667b0e25d4 - <tokio_tungstenite::WebSocketStream<T> as futures::stream::Stream>::poll::hdfde1f9e11380517
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-tungstenite-0.1.2/src/lib.rs:97
  14:     0x55667b0e285a - <futures::stream::split::SplitStream<S> as futures::stream::Stream>::poll::hb4a6de1455b07ea8
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/stream/split.rs:14
  15:     0x55667b097913 - <futures::stream::for_each::ForEach<S, F, U> as futures::future::Future>::poll::hdc98d641d3de2017
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/stream/for_each.rs:45
  16:     0x55667b0e21f3 - <futures::future::map::Map<A, F> as futures::future::Future>::poll::h8cdad3f417638717
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/map.rs:30
  17:     0x55667b0e4516 - <futures::future::map_err::MapErr<A, F> as futures::future::Future>::poll::h8c67f452f112d092
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/map_err.rs:30
  18:     0x55667b0e2efa - <futures::future::select::Select<A, B> as futures::future::Future>::poll::h57c6ee48948d3aac
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/select.rs:48
  19:     0x55667b0ceac5 - <futures::future::chain::Chain<A, B, C>>::poll::h6a75eceb7124d4aa
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/chain.rs:26
  20:     0x55667b0e535c - <futures::future::then::Then<A, B, F> as futures::future::Future>::poll::h5def7d01a5ad2636
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/then.rs:32
  21:     0x55667b13e551 - <alloc::boxed::Box<F> as futures::future::Future>::poll::hb3767344ebe98fd2
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/future/mod.rs:106
  22:     0x55667b127c7c - <futures::task_impl::Spawn<F>>::poll_future::{{closure}}::h48649b1ca7a1bffa
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:337
  23:     0x55667b12807e - <futures::task_impl::Spawn<T>>::enter::{{closure}}::hd669513906380924
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:484
  24:     0x55667b140843 - futures::task_impl::set::{{closure}}::hf10679f8e41b220c
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:61
  25:     0x55667b1298c7 - <std::thread::local::LocalKey<T>>::with::h74b2509b934fff98
                               at /checkout/src/libstd/thread/local.rs:253
  26:     0x55667b1405ef - futures::task_impl::set::h50f076371bd50260
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:54
  27:     0x55667b127fce - <futures::task_impl::Spawn<T>>::enter::hec76d727940c9640
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:484
  28:     0x55667b127c07 - <futures::task_impl::Spawn<F>>::poll_future::hb5a04041cb6e0824
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.12/src/task_impl/mod.rs:337
  29:     0x55667b14c322 - tokio_core::reactor::Core::dispatch_task::{{closure}}::h49b3dd2d7da2a25e
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.6/src/reactor/mod.rs:366
  30:     0x55667b124194 - <scoped_tls::ScopedKey<T>>::set::hdf6c715dc2147ce3
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.0/src/lib.rs:135
  31:     0x55667b14be68 - tokio_core::reactor::Core::dispatch_task::h2ca8f82a3a2b061c
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.6/src/reactor/mod.rs:366
  32:     0x55667b14b3b5 - tokio_core::reactor::Core::dispatch::h0252fa03a28cc3c2
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.6/src/reactor/mod.rs:324
  33:     0x55667b14af1b - tokio_core::reactor::Core::poll::ha5e0c2568b9e79a4
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.6/src/reactor/mod.rs:312
  34:     0x55667b09a2a8 - tokio_core::reactor::Core::run::he9daa5ab6dd0f82d
                               at /home/bbigras/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.6/src/reactor/mod.rs:249
  35:     0x55667b0e5974 - test_tung_server::main::hc8b7e99d2f69155d
                               at /home/bbigras/dev/rust/tests/test-tung-server/src/main.rs:104
  36:     0x55667b18a3f5 - std::panicking::try::do_call::h689a21caeeef92aa
                               at /checkout/src/libcore/ops.rs:2606
                               at /checkout/src/libstd/panicking.rs:454
  37:     0x55667b1915da - __rust_maybe_catch_panic
                               at /checkout/src/libpanic_unwind/lib.rs:98
  38:     0x55667b18ae9a - std::rt::lang_start::hf63d494cb7dd034c
                               at /checkout/src/libstd/panicking.rs:433
                               at /checkout/src/libstd/panic.rs:361
                               at /checkout/src/libstd/rt.rs:57
  39:     0x55667b0e7302 - main
  40:     0x7f506f505510 - __libc_start_main
  41:     0x55667b096159 - _start
  42:                0x0 - <unknown>
```

rustc 1.17.0-nightly (7846dbe0c 2017-03-26)
futures 0.1.12
tokio-core 0.1.6
tokio-tungstenite 0.1.2
tungstenite 0.2.0

**Comment by agalakhov** [comment]
Could you please run cargo test --release on the Tungstenite crate (not tokio-tungstenite) on your machine? This seems to be hardware-dependent due to the recent low-level unsafe optimization.

**Comment by agalakhov** [comment]
Fixed in tungstenite 0.2.1.

**Comment by bbigras** [comment]
> Fixed in tungstenite 0.2.1.

I confirm. Thanks.
